### ISSUE TITLE
Is gun dead?

### ISSUE THREAD

**Issue by obv-mikhail** [amark/gun#627]
There seem to have been very few updates on gun in the past two years. I thought there were full time developers paid to work on gun? Is that not the case? https://techcrunch.com/2018/05/23/gun-raises-more-than-1-5m-for-its-decentralized-database-system/

**Comment by Dletta** [comment]
Mark is the main contributor. I am not aware of any full-time developers.

**Comment by amark** [comment]
@obv-mikhail huh? You haven't been around to see...the past 2 years we've gone from:

 - tens of thousands of ops/sec to tens of millions of ops/sec.

 - 1.5 rewrites from scratch.

 - prototype projects only to production scale with millions of users, top 300 globally ranked sites running us.

Closing.

**Comment by spiffytech** [comment]
Which top-300 sites are using Gun?

**Comment by obv-mikhail** [comment]
@amark Gun looks very impressive - you have done an amazing job with it. But, you haven't answered the main question I had - are there paid developers working on Gun (or are there plans to hire any)?

**Comment by obv-mikhail** [comment]
@amark I want to choose between FeathersJS and Gun for a non P2P project. Gun overall fits my needs better but I can't find any proper examples of how events sent to Gun can be parsed for purposes of rejecting / accepting them based on some token. There have been issues opened here regarding authorization since two years ago, but there still exist no good examples of how non P2P authorization could be implemented.

If Gun doesn't have an active community and isn't actively developed, it would be easier for me to go with Feathers. So, I wanted to find out whether Gun is still active or not.

**Comment by amark** [comment]
@obv-mikhail Our community is about as big as FeathersJS.

What gives you the impression GUN is dead?

https://github.com/zrrrzzt/bullet-catcher

**Comment by projectoblio** [comment]
@obv-mikhail  

You know there's a theory backed by some pretty reputable historians that the world isn't pushed forward by teams of scientists, but by one or two individuals in each generation with the stubbornness / innovative minds to see what other people can't. A salaried employee working on a team is prone to tremendous groupthink and is chained by promises to investors. I would go as far as to say that VC funding only hurts innovation because oftentimes it bogs you down into previous presentations you've already given, instead of looking at other places in tech for inspiration. It's a pretty well known fact in the Valley that working by yourself allows you to innovate and change directions much more quickly than in a team.

You should know that this is one of those "Chaos" patterns that exists at every element of the universe. As a guy with a bio background myself this theory is called `punctual evolution` in biology. There are periods of rapid change where one group of species undergo rapid changes compared to the rest of its history, often caused by one new gene enabling a population burst. I'm sure you would also find this pattern in the areas of fluid dynamics, politics, star constructions, and just about every other studied topic out there.

In short you could not be more wrong about this library. It solves a problem that a lot of people won't appreciate this year but likely will sometime the next decade. Gun does what it's supposed to do; Its DB structure and philosophy aren't for a traditionally monetized project, and its radical change from what the web looks like now compared to what it's going to look like can be hard to grasp.   

So if you know know how to design secure systems using a `token` or something you were taught in school or at another job, you probably won't be able to use this library. If you know what proof-of-work or cryptocurrency is, or been in those spaces as long as I have, using this library is  basically like watching porn.

**Comment by davidbludlow** [comment]
@obv-mikhail You have probably moved on years ago, but I wanted to mention my experience with FeathersJS. I have been using it a while and am looking towards making a system that uses both it an Gun, because I find FeathersJS disappointing, for real-time communication.

If you don't care about FeathersJS, then this post may not be worth your time to read.

Feathers works well for making REST APIs. Feathers says that it can handle real-time collaboration situations, and it can, with great effort, if you want to make it not scalable. The moment that you start making your clients subscribe to real-time updates of data is the moment that you can no longer simply scale by getting Docker Swarm to spin up more containers, each running a copy of your FethersJS app. This is because a FeathersDockerContainer1 will have no way of knowing what changes FeathersDockerContainer2 may or may not be doing to Document1, so FeathersDockerContainer1 may not know when to notify the clients that have a socket.io connection with it that Document1 has changed.

There are workarounds (like [feathers-sync](https://www.npmjs.com/package/feathers-sync) and [feathers-distributed](https://www.npmjs.com/package/@kalisio/feathers-distributed) which were listed on [this list](https://github.com/feathersjs/awesome-feathersjs#scaling)) for feathers, to make it scale a bit more, but all those workarounds are based on the principal of spinning up multiple containers and using a pub/sub service to broadcast every successful DB operation to every other running container. This makes absolutely no sense to me. What is the point of spinning up multiple containers if every container is going to have to process a one network request for each DB update that every other container processes?!?! (If anyone has any evidence against what I said, please tell me, because this glaringly obvious design flaw has been driving me crazy!)

This paragraph will make me sound like I am a sloppy programmer, but I will explain... Two weeks ago, I gave up on using Feathers for anything other than a REST API for our Atlas MongoDB. I instead set up a system where I use [Pusher](pusher.com) (a very expensive pub/sub service) in the after hooks of Feathers to publish little notifications that would basically just say "Document 5edf994287656ada3867582b changed". Since the messages were anonymous, I didn't have to deal with authentication on the Pusher side. Then, clients would use Pusher to subscribe to notifications of when the documents they were interested in changed. When a client received such a notification, then it would fetch an updated copy of the document in question from feathers, with proper authentication, and everything. We spent a long time programming complicated authentication/permission rules in feathers, and I didn't have the time to recreate all those rules with some new system. Also Pusher can only handle very small messages. The system works well, for now, but it was a mistake...

I SHOULD HAVE USED GUN. You see, I thought that Pusher's quota of 200,000 free messages per day would be plenty, but Pusher counts messages in a deceptively harsh way, so we are close to having to pay a bunch of money per month. I should have used Gun with a super-easy-to-set-up $7/month relay server deployed on Heroku (I didn't like how the free Heroku servers fell asleep with inactivity), That cheap relay server, on its own, with default settings, can handle over 2,000 gun messages per second, according to my tests (and I bet it could do much more, if I bothered to try).

Anyways, @obv-mikhail , it sounds like you probably chose to use FeathersJS instead of Gun like 4 years ago. Lol. If that is true, then perhaps you could work with me to explore ways to use Gun with feathers ☺ I cannot rewrite my entire backed to use Gun. My company signed multiple contracts that place strict rules on how my website can store my clients' private data, but there are some things I can use Gun for. If we wrote an adapter to use gun for real-time communication between FeathersJS containers, (I mean, there is already [feathers-distrubuted](https://www.npmjs.com/package/@kalisio/feathers-distributed) which sorta maybe does something similar to that, using cote instead of Gun, but I don't know if I can trust it, because I am 75% sure it just pretends to solve the problem by making every feathers docker container be notified of every service event, which sounds inherently not scalable.)

Note that FeathersJS's system of channels and subscribing is abhorrently complex. Even if you were to solve the scaling issue, feather's complex system of channels and subscribing makes it difficult to do anything but notify a user of all the activity that every other user is doing for a given service, which is almost never something that needs to be done in real-life applications. (In most real-life applications, that activity should be filtered to only notify users of the activity that that user is interested in.)

**Comment by backspaces** [comment]
Here is an interesting way to see how active  and up to date gun is.
- Can you use es6 import with it?
- Can deno import it (using es6 imports)

Note that deno is striving to have the node api available for those porting from node to deno.
